### ISSUE TITLE
data science feedback

### ISSUE THREAD

**Issue by temphstarmans** [ElektorLabs/esp32-energymeter#7]
There are trivial applications one could do with the energy meter data like computing energy consumption.
Less trivial is to detect whether something strange occurs. Obviously you can provide a lot of value to your customer by preventing; production of invalid parts, solar roof is burning, electricity shortage etc. In the following, I would like to briefly outline how this can be done.
I will discuss two approaches; Hidden Markov Models and SPC charts. In the ideal factory situation, your energy monitor would be detected to one say injection molding machine. I guess it is harder if you look at all the machines at the same time.

# Hidden Markov Models

Look at the following example, [Dishonest Casino ](https://hmmlearn.readthedocs.io/en/latest/auto_examples/plot_casino.html), this outlines how you could spot by just looking at the dice rolls whether the dice is loaded or not. Just as casinos analyze die roll patterns to identify irregularities, we can examine the data produced by the energy sensor to pinpoint if our machine is not working correctly. 
Basically the procedure is as follows. You collect data from your sensors along time. You resample to say 1 second by taking the mean. Resampling speeds up the calculation. You split your data in a train and test set. For simplicity, I ignore the validation set. You could further simplify by using an one dimensional treatment.
You then fit a Gaussian Hidden Markov upon your data. Your data is not discrete but continuous. You try models with different number of states.   Each models will have a different goodness of fit using the AIC, BIC or LLS criterion. By looking at the the metrics and whether or not the test set delivers the same result as the train set you can see if you find something. Let's say you try models from 2 up to 10 states and see that the best fit is around 4.You then plot the state number along your data. You plot how often the system resides in the state, e.g. by using a fraction. You should then be able to label something like; off, normal, startup and error. 

# Statistical Process Control

Another approach would be to use [Statistical Process Control](https://en.wikipedia.org/wiki/Statistical_process_control).
A good intro is [here](https://www.sheffieldmca.org.uk/UserFiles/File/Control_Charts_One_Page_Book_Revised_June_14.pdf) basically you need a package which implements several rules; e.g. outside 3 sigma boundary is outlier point.
Several packages are available [pyspc](https://github.com/carlosqsilva/pyspc) and [statprocon](https://github.com/mattmccormick/statprocon). They seem to be incomplete.
